### PR TITLE
Performance Updates

### PR DESCRIPTION
**Pull request by aroc**
## What does this PR do? How does it affect users?
It makes some changes to the codebase to increase performance, specifically when dealing with large responses / large query result datasets.

The changes it makes:
- No longer using `_.cloneDeep` everywhere we make updates. This cloning is very costly. Instead, we only pass in the properties we need to update and use a `_.merge` call. We also *never* clone the response object and instead just assign the new model's `response` reference to the old model's `response` object.
- We no longer initialize a new `Dataviz` object with the entire response object each time we want to check which chart types are supported for the given query. We borrowed some code from Keen.js and put it into this project for *much* faster checking of what chart types are supported.

## How should this be tested?

1) Try running some queries with large results, especially ones with a large number of groups.

Step through the code line by line. Things to keep in mind as you review:
 - Are there any edge cases not covered by this code?
 - Does this code follow conventions (naming, formatting, modularization, etc) where applicable?

Fetch the branch and/or deploy to staging to test the following:

- [x] Does the code compile without warnings (check shell, console)?
- [x] Do all tests pass?
- [x] Does the UI, pixel by pixel, look exactly as expected (check various screen sizes, including mobile)?
- [x] If the feature makes requests from the browser, inspect them in the Web Inspector. Do they look as expected (parameters, headers, etc)?
- [x] If the feature sends data to Keen, is the data visible in the project if you run an extraction (include link to collection/query)?
- [ ] If the feature saves data to a database, can you confirm the data is indeed created in the database?

## Related tickets?
N/A

### CODE REVIEW

File: client/js/app/actions/ExplorerActions.js [hunk 1]
```diff
@@ -217,7 +217,8 @@ var ExplorerActions = {
     updates.loading = false;
 
     if (!ExplorerUtils.responseSupportsChartType(response, explorer.metadata.visualization.chart_type, explorer.query.analysis_type)) {
-      updates.metadata.visualization.chart_type = ExplorerUtils.getChartTypeOptions(response, explorer.query.analysis_type)[0];
+      updates.metadata = _.cloneDeep(explorer.metadata);
+      updates.metadata.visualization.chart_type = ExplorerUtils.getChartTypeOptions(response.query)[0];
```
Comment (joannecheng) [inline client/js/app/actions/ExplorerActions.js]: Getting this error when I run a `count` query: `Uncaught TypeError: Cannot read property '0' of undefined`

Comment (joannecheng) [inline client/js/app/actions/ExplorerActions.js]: project id: 5011efa95f546f2ce2000000
count on `pageviews`

File: client/js/app/utils/ExplorerUtils.js [hunk 2]
```diff
@@ -210,32 +207,74 @@ module.exports = {
...
+    if (query.analysis_type === "select_unique") {
+      dataType = 'nominal';
+    }
+
+    return dataType;
+  },
+
+  getChartTypeOptions: function(query) {
+    var dataTypes = {
+      'categorical':        ['piechart', 'barchart', 'columnchart', 'table'],
```
Comment (joannecheng) [inline client/js/app/utils/ExplorerUtils.js]: it looks like `singular` is missing from the `dataTypes` object in `getChartTypeOptions` in ExporerUtils

Comment (joannecheng) [inline client/js/app/utils/ExplorerUtils.js]: adding `'singular':           ['metric']` to the object seems to fix it

File: client/js/app/actions/ExplorerActions.js (hunk 1 above)
Comment (aroc) [inline client/js/app/actions/ExplorerActions.js]: Looks like the `singular` data type addition fixed this. Can you take another look when you get a moment?

**Comment by joannecheng** [comment]
:+1: everything looks good now.
